pkg/collector/hostname_collector.go

```diff
@@ -0,0 +1,119 @@
+package collector
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	//autoscalingv2 "k8s.io/api/autoscaling/v2beta2"
+	autoscalingv2 "k8s.io/api/autoscaling/v2"
+)
+
+const (
+	HostnameMetricType = "requests-per-second"
+	HostnameRPSQuery   = `scalar(sum(rate(%s{host=~"%s"}[1m])) * %.4f)`
+)
+
+type HostnameCollectorPlugin struct {
+	metricName string
+	promPlugin CollectorPlugin
+}
+
+type HostnameCollector struct {
+	interval      time.Duration
+	promCollector Collector
+}
+
+func NewHostnameCollectorPlugin(
+	promPlugin CollectorPlugin,
+	metricName string,
+) (*HostnameCollectorPlugin, error) {
+	if metricName == "" {
+		return nil, fmt.Errorf("Failed to initialize hostname collector plugin, metric name was not defined")
+	}
+
+	return &HostnameCollectorPlugin{
+		metricName: metricName,
+		promPlugin: promPlugin,
+	}, nil
+}
+
+// NewCollector initializes a new skipper collector from the specified HPA.
+func (p *HostnameCollectorPlugin) NewCollector(
+	hpa *autoscalingv2.HorizontalPodAutoscaler,
+	config *MetricConfig,
+	interval time.Duration,
+) (Collector, error) {
+	if config == nil {
+		return nil, fmt.Errorf("Metric config not present, it is not possible to initialize the collector.")
+	}
+	// Need to copy config and add a promQL query in order to get
+	// RPS data from a specific hostname from prometheus. The idea
+	// of the copy is to not modify the original config struct.
+	confCopy := *config
+
+	if _, ok := config.Config["hostnames"]; !ok {
+		return nil, fmt.Errorf("Hostname is not specified, unable to create collector")
+	}
+	regex, err := regexp.Compile("^[a-zA-Z0-9.-]+$")
```

Comment:
nit: I think we have compile this on as a package global variable and avoid compiling it for every initialization.

Reply:
Oh, cool, I'll check it

Reply:
I don't see what you describe reflected in the code. Are you just proposing it?

I think the change of it being misused if it's a package local variable is very low. It's essentially used as a constant, but can't be constant because of the type.

Initializing it in the factory seems ok as it would only initialize it once, but you still need to pass it on plugin initialization which is maybe a bit overkill considering the _constant_ nature of the variable. My main point was to not compile the regex multiple times for no reason as it's not dynamic in any way.